draft-ietf-dnsop-svcb-https.md

```diff
@@ -736,20 +720,31 @@ support, and this informs the underlying transport protocol used (such
 as QUIC-over-UDP or TLS-over-TCP).
 
 ALPNs are identified by their registered "Identification Sequence"
-(alpn-id), which is a sequence of 1-255 octets.
+(`alpn-id`), which is a sequence of 1-255 octets.
+
+    alpn-id = 1*255OCTET
+
+The presentation `value` of "alpn" is a comma-separated list of one or
+more `alpn-id`s, each represented as a character-string:
+
+    alpn-value = char-string *("," char-string)
+
+To decode an `alpn-value`, the parser MUST split the `value` at each ","
+character, and then apply the character-string decoding algorithm to each
+segment.  (See {{decoding}} for the definition of `char-string`.)  Accordingly,
+literal commas cannot be used in these `char-string`s, and MUST be represented
+by their decimal escape sequence ("\044") instead.
+Thus, a valid `pair` for the "alpn" key could look like
 
-    alpn-id = 1*255(OCTET)
+    alpn="example1,\"weird\\044 why?\",example2"
 
-The presentation value of "alpn" is a comma-separated list of one or
-more `alpn-id`s.  Any commas present in the protocol-id are escaped
-by a backslash:
+This would be decoded as three `alpn-id`s: "example1", "weird, why?", and
```

Comment:
The double quotes are part of the second alpn-id. so this should read.

This would be decoded as three `alpn-id`s: 'example1', '"weird, why?"', and
'example2'.

Reply:
In this PR, this example is first parsed to a sequence of three `char-string`s: `example1`, `"weird\044 why?"`, and `example2`.  Then the character-string decoding algorithm is applied to each `char-string`, yielding `example1`, `weird, why?`, and `example2`.  This is because double-quotes are optional in `char-string`.

My goal here is to make the parsing algorithm easy to implement, so please suggest changes.  I proposed this formulation to minimize the amount of new parsing code required (reuse the same decoding algorithm), while making sure that all `alpn-id`s are representable, including those that contain commas and spaces.

Reply:
Detecting escaped commas is dead easy.  GO LOOK AT HOW BIND DOES IT.  All this just makes it harder for the person entering the data.

Reply:
Also the point of using quoted strings is to turn off special processing to get the full token at places that would otherwise terminate the token.   Once you have the full token you don't need to drop them a second time.

Reply:
And so you don't have to search around in the repository to find it. 

At this point the double quotes have been processed.  The token still contains all the escapes.

```
+static isc_result_t
+alpn_fromtxt(isc_textregion_t *source, isc_buffer_t *target) {
+       isc_textregion_t source0 = *source;
+       do {
+               RETERR(commatxt_fromtext(&source0, true, target));
+       } while (source0.length != 0);
+       return (ISC_R_SUCCESS);
+}
```

```
@@ -1390,7 +1432,8 @@ name_length(const dns_name_t *name) {
 }
 
 static isc_result_t
-txt_totext(isc_region_t *source, bool quote, isc_buffer_t *target) {
+commatxt_totext(isc_region_t *source, bool quote, bool comma,
+               isc_buffer_t *target) {
        unsigned int tl;
        unsigned int n;
        unsigned char *sp;
@@ -1435,10 +1478,12 @@ txt_totext(isc_region_t *source, bool quote, isc_buffer_t *target) {
                /*
                 * Escape double quote and backslash.  If we are not
                 * enclosing the string in double quotes also escape
-                * at sign and semicolon.
+                * at sign and semicolon unless comma is set.  If
+                * comma is set then escape commas.
                 */
-               if (*sp == 0x22 || *sp == 0x5c ||
-                   (!quote && (*sp == 0x40 || *sp == 0x3b))) {
+               if (*sp == 0x22 || *sp == 0x5c || (comma && *sp == 0x2c) ||
+                   (!comma && !quote && (*sp == 0x40 || *sp == 0x3b)))
+               {
                        if (tl < 2) {
                                return (ISC_R_NOSPACE);
                        }
@@ -1465,9 +1510,14 @@ txt_totext(isc_region_t *source, bool quote, isc_buffer_t *target) {
 }
 
 static isc_result_t
-txt_fromtext(isc_textregion_t *source, isc_buffer_t *target) {
+txt_totext(isc_region_t *source, bool quote, isc_buffer_t *target) {
+       return (commatxt_totext(source, quote, false, target));
+}
+
+static isc_result_t
+commatxt_fromtext(isc_textregion_t *source, bool comma, isc_buffer_t *target) {
        isc_region_t tregion;
-       bool escape;
+       bool escape, seen_comma = false;
        unsigned int n, nrem;
        char *s;
        unsigned char *t;
@@ -1496,6 +1546,10 @@ txt_fromtext(isc_textregion_t *source, isc_buffer_t *target) {
        }
        while (n-- != 0) {
                c = (*s++) & 0xff;
+               if (comma && !escape && c == ',') {
+                       seen_comma = true;
+                       break;
+               }
                if (escape && (d = decvalue((char)c)) != -1) {
                        c = d;
                        if (n == 0) {
@@ -1534,11 +1588,32 @@ txt_fromtext(isc_textregion_t *source, isc_buffer_t *target) {
        if (escape) {
                return (DNS_R_SYNTAX);
        }
+       if (comma) {
+               /*
+                * Disallow empty ALPN at start or in the middle.
+                */
+               if (s == source->base || (seen_comma && s == source->base + 1))
+               {
+                       return (DNS_R_SYNTAX);
+               }
+               isc_textregion_consume(source, s - source->base);
+               /*
+                * Disallow empty ALPN at end.
+                */
+               if (seen_comma && source->length == 0) {
+                       return (DNS_R_SYNTAX);
+               }
+       }
        *tregion.base = (unsigned char)(t - tregion.base - 1);
        isc_buffer_add(target, *tregion.base + 1);
        return (ISC_R_SUCCESS);
 }

 +static isc_result_t
+txt_fromtext(isc_textregion_t *source, isc_buffer_t *target) {
+       return (commatxt_fromtext(source, false, target));
+}
+
 static isc_result_t
 txt_fromwire(isc_buffer_t *source, isc_buffer_t *target) {
        unsigned int n;
```

Reply:
OK, I've updated the PR to reflect this approach.  Please review.

Reply:
I would say that they are key value pairs.  If the definition of a value consists of a list (i.e. ipv4hints, ipv6hints, alpn) the list elements are separated by a unescaped ",".  To add a literal comma to an element in a list, the comma must be escaped.   At the moment you have everything defined to be a list including the generic form which clearly is not a list.

Reply:
OK, I've changed the text to reflect this idea.